src/lua/protobuf_wireformat.lua

```diff
@@ -64,6 +82,32 @@ local function as_int64(value)
     return ffi.string(ffi.cast('char *', p), 8)
 end
 
+-- Number representation of 64-bit two's complement value.
+local function from_int64(value)
+    return ffi.cast('uint64_t *', value)[0]
+end
+
+-- Converts input unsigned cdata into signed cdata or lua number if
+-- it fits lua number integer representation without rounding range.
+local function cast_signed_type(value)
+    assert(type(value) == 'cdata')
+    value = ffi.cast('int64_t', value)
+    if value < 2LL^53 - 1 and value > -2LL^53 + 1 then
+        value = tonumber(value)
+    end
+    return value
+end
+
+-- Converts input unsigned cdata into lua number if it fits lua
+-- number integer representation without rounding range.
+local function cast_unsigned_type(value)
+    assert(type(value) == 'cdata')
+    if value < 2ULL^53 - 1 then
+        value = tonumber(value)
+    end
+    return value
+end
```

Comment:
In the RFC we discussed that values of the `number` type are produced for the `[-10^14+1; 10^14-1]` range and `cdata<uint64_t>` or `cdata<int64_t>` for values outside (and uint64_t vs int64_t choose is based on the schema type).

Here we violate this agreement about the ranges from the RFC. The motivation was to be consistent with other tarantool functions like `msgpack.decode`.

I had a guess that we can achieve it by using some already existing function like `tonumber64`, but it seems it works this way only for a string input (not cdata).

I've looked into `src/lua/msgpack.ffi.lua` and it seems that it performs it manually (see the `decode_u64` and `decode_i64` functions). OK. Let's at least steal the constants from it.

Reply:
Added constants and logic identical to `src/lua/msgpack.ffi.lua`.